tests/integration/test_charm.py

```diff
@@ -37,24 +42,248 @@ async def test_build_and_deploy(ops_test: OpsTest):
+    """Test that data is preserved during scale up and scale down."""
+    # Insert values into test table from the primary unit
+    random_unit = ops_test.model.applications[APP_NAME].units[0]
+    server_config_credentials = await get_server_config_credentials(random_unit)
+
+    primary_unit = await get_primary_unit(
+        ops_test,
+        random_unit,
+        server_config_credentials["username"],
+        server_config_credentials["password"],
+    )
+
+    random_chars = generate_random_string(40)
+    create_records_sql = (
+        "CREATE DATABASE IF NOT EXISTS test",
+        "CREATE TABLE IF NOT EXISTS test.instance_state_replication (id varchar(40), primary key(id))",
+        f"INSERT INTO test.instance_state_replication VALUES ('{random_chars}')",
+    )
+
+    mysqlsh_sql_base_command = f"mysqlsh --sql {server_config_credentials['username']}:{server_config_credentials['password']}@127.0.0.1"
+
+    await run_command_on_unit(
+        primary_unit, f"{mysqlsh_sql_base_command} -e \"{';'.join(create_records_sql)}\""
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+    old_unit_names = [unit.name for unit in ops_test.model.applications[APP_NAME].units]
+
+    # Add a unit and wait until it is active
+    await ops_test.model.applications[APP_NAME].add_unit()
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 4
+
+    added_unit = [
+        unit
+        for unit in ops_test.model.applications[APP_NAME].units
+        if unit.name not in old_unit_names
+    ][0]
+
+    # Ensure that all units have the above inserted data
+    for unit in ops_test.model.applications[APP_NAME].units:
+        output = await run_command_on_unit(
+            unit,
+            f"{mysqlsh_sql_base_command} -e \"SELECT * FROM test.instance_state_replication WHERE id = '{random_chars}'\"",
+        )
+        assert random_chars in output
+
+    # Destroy the recently created unit and wait until the application is active
+    await ops_test.model.destroy_units(added_unit.name)
+    time.sleep(5)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        raise_on_blocked=True,
+        timeout=1000,
+    )
+
+    assert len(ops_test.model.applications[APP_NAME].units) == 3
+
+    # Ensure that the data still exists in all the units
+    for unit in ops_test.model.applications[APP_NAME].units:
+        output = await run_command_on_unit(
+            unit,
+            f"{mysqlsh_sql_base_command} -e \"SELECT * FROM test.instance_state_replication WHERE id = '{random_chars}'\"",
+        )
+        assert random_chars in output
+
+    await ops_test.model.set_config({"update-status-hook-interval": "60m"})
+
+
+async def get_primary_unit(
```

Comment:
Can this be moved to tests/integration/helpers.py?

Reply:
moved in ff877d16df482a9781067a33b0a1cf1e46591782